### ISSUE TITLE
Can not execute simplest dll-based invocation

### ISSUE THREAD

**Issue by fresh-ji** [ADLINK-IST/opensplice#108]
Hi,
recently i have been learning the isocpp2-API-based examples in PXXX-VortexOpenSplice-6.9.181018OSS-HDE-x86_64.win-vs2017-installer. I built my own dll-based application, but the dds-related sentences can not be executed. Full codes provided below:

![捕获1](https://user-images.githubusercontent.com/25540984/69770372-c1722680-11c3-11ea-8925-4d1637c3e367.PNG)
dll header

![捕获2](https://user-images.githubusercontent.com/25540984/69770182-1a8d8a80-11c3-11ea-8920-70daadf8c373.PNG)
dll source

![捕获3](https://user-images.githubusercontent.com/25540984/69770191-1fead500-11c3-11ea-8c4d-ca7baaf2ea69.PNG)
main

everytime the program enters dll corretly, when executing "dds::domain::DomainParticipant dp(org::opensplice::domain::default_id());", the cmd window pauses for a few seconds and shuts itself down, nothing else happen.

sorry to disturb you, but i am very confused about this.

thanks.

**Comment by ThijsSassen** [comment]
@fresh-ji looking at your code this would be expected behavior. You only create a domain allocated to a local variable dp and then your function returns resulting in the domain running out of scope and terminating itself. Please look at the [ Isocpp2 documentation](http://download.ist.adlinktech.com/docs/Vortex/apis/ospl/isocpp2/html/index.html) for more information on how it works.

**Comment by fresh-ji** [comment]
i am sorry that i did not express myself clearly, but the application really terminates unnormally.
![捕获2](https://user-images.githubusercontent.com/25540984/69792057-c5706980-1200-11ea-90a4-c1c651699858.PNG)

![捕获1](https://user-images.githubusercontent.com/25540984/69792066-c99c8700-1200-11ea-8c9d-56ae78b949a9.PNG)

i executed it in PoerShell. You can see that "that point" did not printed.

Today i reconsidered it again, i noticed that "#include <Windows.h>" may be the reason, my ide is vs2013. is this a proper way to invoke dds-api in dll?

thanks!

**Comment by ThijsSassen** [comment]
@fresh-ji I think what you seeing is the termination behavior of OpenSplice. When the domain participant runs out of scope it will be cleaned up and when the last domain participant is gone the application will be terminated with an exit call resulting in termination of your application. To prevent this you could keep a global reference to the dp variable resulting in OpenSplice not terminating.

**Comment by fresh-ji** [comment]
But in fact i did not do anything else in my application, whatever i write behind "dds::domain::DomainParticipant dp(org::opensplice::domain::default_id());" can not be reached. So i think the domain participant could not run out of scope.

i looked over #99 [https://github.com/ADLINK-IST/opensplice/issues/99](url) and i think the problems may be the same. 

when i doing this process not in dll-way, everything goes right; and i add  "#include <Windows.h>" to it, the same situation happens; so i think the problem is Windows SDK or something.

Now if i want to package the dds-api into dll and invoke it in pure-C environment, what is the right way to do without using "#include <Windows.h>"?

**Comment by ThijsSassen** [comment]
@fresh-ji To compile an isocpp2 application on windows please see [this](https://istkb.adlinktech.com/article/isoc2-vortex-opensplicedds-based-applications/) One of the steps is that any include file from OpenSplice should be done before including Windows.h to avoid memory heap crashes.

**Comment by fresh-ji** [comment]
very appreciate for your answer. putting Windows.h backwards could make the non-dll application run correctly. but i use "LoadLibrary" to load dll, which do not use any include file from OpenSplice, only Windows.h, so the problem is still exists. i am going to reconsider the framework to avoid it. Thanks!
Here is my anothor question. when i run my another application, the publisher writes messages every one second, the subscriber takes them and shows in console, for total 19 messages. But everytime the subscriber receives the first message twp times("initederate"), you can see from the figure, at 0.00 and 0.06. 
i am for sure my "initederate" sends only once, how the second message been received.
at last of figure, the "letusendit" also receives twice, which was sent only once.
thanks for looking through this.

**Comment by ThijsSassen** [comment]
@fresh-ji what code are you using to read the samples? Are you using read or take?

**Comment by fresh-ji** [comment]
i am using “take”

![捕获1](https://user-images.githubusercontent.com/25540984/69836873-c1872a80-1286-11ea-921a-28b3f013ab1a.PNG)

![捕获2](https://user-images.githubusercontent.com/25540984/69836875-c6e47500-1286-11ea-9992-2477b9058426.PNG)

**Comment by fresh-ji** [comment]
@Johnnyxy at https://github.com/ADLINK-IST/opensplice/issues/99 has solved my issue perfectly. Thanks  for everyone.
